lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py

```diff
@@ -1,31 +1,106 @@
 import pytest
 
 
-def test_pbench_list_tools_help():
-    command = ["pbench-list-tools", "--help"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert b"Usage: pbench-list-tools [OPTIONS]" in out
-    assert exitcode == 0
-
-
-def test_list_tool(monkeypatch, agent_config, pbench_run, pbench_cfg):
-    monkeypatch.setenv("_PBENCH_AGENT_CONFIG", str(pbench_cfg))
-    p = pbench_run / "tools-v1-default" / "testhost.example.com"
-    p.mkdir(parents=True)
-    tool = p / "perf"
-    tool.touch()
-
-    command = ["pbench-list-tools"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"default: testhost.example.com ['perf']" in out
-
-    command = ["pbench-list-tools", "--group", "default"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"default: testhost.example.com ['perf']" in out
-
-    command = ["pbench-list-tools", "-n", "perf"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"tool name: perf groups: default" in out
+EMPTY = b""
+USAGE = b"Usage: pbench-list-tools [OPTIONS]"
+BAD_GROUP = b"Bad tool group: "
+TRACEBACK = b"Traceback (most recent call last):\n"
+
+
+class Test_list_tools_no_tools_registered:
+    def test_help(self):
+        command = ["pbench-list-tools", "--help"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert TRACEBACK not in err
+        assert EMPTY == err
+        assert USAGE in out
+        assert exitcode == 0
+
+    def test_no_args(self, agent_config):
+        command = ["pbench-list-tools"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert TRACEBACK not in err
+        assert EMPTY == err
+        assert EMPTY == out
+        assert exitcode == 0
+
+
+class Test_list_tools_tools_registered:
+    @pytest.fixture
+    def tool(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        tool = p / "perf"
+        tool.touch()
+
+    @pytest.fixture
+    def tool_with_random_file(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        (p.parent / "foo").touch()
+        tool = p / "perf"
+        tool.touch()
```

Comment:
Not sure yet, but probably - see your comment below.

Reply:
I could go either way on this.  Having this setup encapsulated is a feature, especially since this fixture closely follows the one above it.  On the other hand, there is a cost to forcing the reader to go somewhere else to find everything.  But, I think the encapsulation is probably worth the cost.

Reply:
I don't know about other IDEs (maybe I'm in the slums and nobody else has this issue), but vscode doesn't see the fixture "parameters" as references (not really surprising, since it'd take special pytest awareness to recognize this), which makes tracking the use or origin of fixtures awkward. It's a great feature, if extraordinarily quirky; but overuse just makes the code harder to read. Still, I'm not objecting, just questioning; so I'll resolve this conversation and let it go. 😁

Reply:
Yeah, PyCharm doesn't like them much, either.

Of course, there's no huge advantage to this being a "fixture" over being a "helper routine"...the difference is that, as a fixture, it's called implicitly, in exchange for the weird presence in the function signature.

So long as it's not using any of the other superpowers of a fixture (like implicit invocations at broader scopes, clean-up actions, et al.), it could just as easily be a simple function call, and perhaps that would be better.  (But, as you say, this conversation has been resolved, so....)